### PR TITLE
UX-1059, UX-960/Selectors performance issues

### PR DESCRIPTION
**Pull request by xantorres**
## Performance enhancements + new dashboard
____

#### Selectors:

- the default selector would return a new array when no data existed which caused selectors to re-run constantly. 
- fixed O(n^2) comprehensions in selectors
- completely revamped combined hosts calculations to be a single scan vs 8 scans.

#### Dashboard

- Limited number of dashboard cards that are rendered to be
    - clusters
    - nodes
    - clouds
- when in a decco environment you can customize the dashboard
    - drag and drop reorder
    - click to toggle add / remove

#### Default dashboard view - traditional
image

#### Default dashboard view - decco
image

#### Edit dashboard view - decco 
image

### CODE REVIEW

File: src/app/plugins/kubernetes/components/dashboard/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    // remove the add action when not permitted to
+    return report.permissions.includes(role) ? report : { ...report, addRoute: '' }
+  })
+  const filteredReports = mappedReports.filter((report) => {
+    if (!report.overallPermissions) {
+      return report
+    }
+    return report.overallPermissions.includes(role)
+  })
+  // if (cardOrder?.length > 0) {
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/dashboard/helpers.ts:63]: is this still needed?

File: src/app/plugins/kubernetes/components/infrastructure/common/combineHosts.ts [hunk 2]
```diff
@@ -1,47 +1,88 @@
...
+const getCloudStack = (stackSet) => {
+  if (stackSet.size === 2) {
+    return CloudStack.Both
+  }
+  if (stackSet.size === 1) {
+    return stackSet.values().next().value
+  }
+  return CloudStack.Unknown
+}
+const formattedRoleMapping = {
```
Comment (kevinylee1991) [inline src/app/plugins/kubernetes/components/infrastructure/common/combineHosts.ts:27]: just a thought, we could deprecate a lot of the openstack stuff

**Review by kevinylee1991 — APPROVED** [review]

**Review by jruddell — APPROVED** [review]

**Review by knnguy — APPROVED** [review]

File: src/app/core/caching/cacheReducers.ts [hunk 3]
```diff
@@ -94,8 +94,9 @@ const reducers = {
     }>,
   ) => {
     const dataLens = lensPath([dataStoreKey, cacheKey])
+    const result = over(dataLens, append(mergeLeft(params, item)))(state)
 
-    return over(dataLens, append(mergeLeft(params, item)))(state)
+    return result
```
Comment (xantorres) [inline src/app/core/caching/cacheReducers.ts:99]: why creating a useless and pointless variable here?

File: src/app/rootReducer.ts [hunk 4]
```diff
@@ -11,7 +16,7 @@ import themeReducers, { customThemeKey } from 'core/session/themeReducers'
 const persistCacheConfig = {
   key: cacheStoreKey,
   storage,
-  blacklist: ['loadingData'],
+  blacklist: [loadingStoreKey, dataStoreKey, updatingStoreKey],
```
Comment (xantorres) [inline src/app/rootReducer.ts:19]: What is `updatingStoreKey` and why has it been added?

**Pull request merged**